src/Data/Text/Internal/Encoding.hs

```diff
@@ -0,0 +1,522 @@
+-- indices to let us avoid this function.
+slowValidateUtf8ChunkFrom :: Int -> ByteString -> (Int, Maybe Utf8State)
+slowValidateUtf8ChunkFrom ofs bs = loop ofs ofs utf8StartState
+  where
+    loop !utf8End i s
+      | i < B.length bs =
+          case updateUtf8State (B.index bs i) s of
+            Just s' ->
+              let utf8End' = if isUtf8StateIsComplete s' then i + 1 else utf8End
+              in loop utf8End' (i + 1) s'
+            Nothing -> (utf8End, Nothing)
+      | otherwise = (utf8End, Just (Utf8State s (partUtf8UnsafeAppend partUtf8Empty (B.drop utf8End bs))))
+
+-- | Validate another 'ByteString' chunk in an ongoing stream of UTF-8-encoded text.
+--
+-- Returns a pair:
+--
+-- 1. The first component @n@ is the end position, relative to the current
+--    chunk, of the longest prefix of the accumulated bytestring which is valid UTF-8.
+--    @n@ may be negative: that happens when an incomplete code point started in
+--    a previous chunk and is not completed by the current chunk (either
+--    that code point is still incomplete, or it is broken by an invalid byte).
+--
+-- 2. The second component @ms@ indicates the following:
+--
+--     - if @ms = Nothing@, the remainder of the chunk contains an invalid byte,
+--       within four bytes from position @n@;
+--     - if @ms = Just s'@, you can carry on validating another chunk
+--       by calling 'validateUtf8More' with the new state @s'@.
+--
+-- === Properties
+--
+-- Given:
+--
+-- @
+-- 'validateUtf8More' s chunk = (n, ms)
+-- @
+--
+-- - If the chunk is invalid, it cannot be extended to be valid.
+--
+--     @
+--     ms = Nothing
+--     ==> 'validateUtf8More' s (chunk '<>' more) = (n, Nothing)
+--     @
+--
+-- - Validating two chunks sequentially is the same as validating them
+--   together at once:
+--
+--     @
+--     ms = Just s'
+--     ==> 'validateUtf8More' s (chunk '<>' more) = 'Data.Bifunctor.first' ('Data.ByteString.length' chunk '+') ('validateUtf8More' s' more)
+--     @
+--
+-- @since 2.0.2
+validateUtf8More :: Utf8State -> ByteString -> (Int, Maybe Utf8State)
+validateUtf8More st@(Utf8State s0 part) bs
+  | len > 0 = loop 0 s0
+  | otherwise = (- partUtf8Len part, Just st)
+  where
+    len = B.length bs
+    -- Complete an incomplete code point (if there is one)
+    -- and then jump to validateUtf8ChunkFrom
+    loop !i s
+      | isUtf8StateIsComplete s = validateUtf8ChunkFrom i bs
+      | i < len =
+        case updateUtf8State (B.index bs i) s of
+          Nothing -> (- partUtf8Len part, Nothing)
+          Just s' -> loop (i + 1) s'
+      | otherwise = (- partUtf8Len part, Just (Utf8State s (partUtf8UnsafeAppend part bs)))
+
+-- | A delayed representation of strict 'Text'.
+
+-- For internal purposes, this is instead used as a delayed 'Array':
+-- it may not actually be valid 'Text' (e.g., 'word8ToStrictBuilder',
+-- 'byteStringToStrictBuilder', 'partUtf8ToStrictBuilder').
+--
+-- @since 2.0.2
+data StrictBuilder = StrictBuilder
```

Comment:
Let's move this to a separate internal, un-exposed module. 

I'm worried that this structure might be too tempting for users to use this strict builder outside of encoding, even while we consider it a very internal thing. Maybe add a strict warning discouraging such usage?

Reply:
I changed the names of the functions that break this to `unsafeFromByteString` and `unsafeFromWord8`.